src/coreclr/src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,870 @@
+
+    DWORD cbSig;
+    PCCOR_SIGNATURE pSig = AllocateSignature(info.LoaderAllocator, sigBuilder, &cbSig);
+
+    SigTypeContext emptyCtx;
+
+    ILStubLinker sl(info.Caller->GetModule(),
+                    Signature(pSig, cbSig),
+                    &emptyCtx,
+                    NULL,
+                    FALSE,
+                    FALSE);
+
+    ILCodeStream* pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+
+    // void CallTarget(void* argBuffer, void* retVal, void** pTailCallAwareRetAddress)
+    const int ARG_ARG_BUFFER = 0;
+    const int ARG_RET_VAL = 1;
+    const int ARG_PTR_TAILCALL_AWARE_RET_ADDR = 2;
+
+    auto emitOffs = [&](UINT offs)
+    {
+        pCode->EmitLDARG(ARG_ARG_BUFFER);
+        if (offs != 0)
+        {
+            pCode->EmitLDC(offs);
+            pCode->EmitADD();
+        }
+    };
+
+    StackSArray<DWORD> argLocals;
+    for (COUNT_T i = 0; i < info.ArgBufLayout.Values.GetCount(); i++)
+    {
+        const ArgBufferValue& arg = info.ArgBufLayout.Values[i];
+        DWORD argLcl = pCode->NewLocal(LocalDesc(arg.TyHnd));
+        argLocals.Append(argLcl);
+
+        // arg = args->Arg_i
+        emitOffs(arg.Offset);
+        EmitLoadTyHnd(pCode, arg.TyHnd);
+        pCode->EmitSTLOC(argLcl);
+    }
+
+    DWORD targetAddrLcl;
+    if (info.ArgBufLayout.HasTargetAddress)
+    {
+        targetAddrLcl = pCode->NewLocal(ELEMENT_TYPE_I);
+
+        emitOffs(info.ArgBufLayout.TargetAddressOffset);
+        pCode->EmitLDIND_I();
+        pCode->EmitSTLOC(targetAddrLcl);
+    }
+
+    // RuntimeHelpers.FreeTailCallArgBuffer();
+    pCode->EmitCALL(METHOD__RUNTIME_HELPERS__FREE_TAILCALL_ARG_BUFFER, 0, 0);
+
+    // *pTailCallAwareRetAddr = NextCallReturnAddress();
+    pCode->EmitLDARG(ARG_PTR_TAILCALL_AWARE_RET_ADDR);
+    pCode->EmitCALL(METHOD__STUBHELPERS__NEXT_CALL_RETURN_ADDRESS, 0, 1);
+    pCode->EmitSTIND_I();
+
+    int numRetVals = info.CallSiteSig->IsReturnTypeVoid() ? 0 : 1;
+    // Normally there will not be any target and we just emit a normal
+    // call/callvirt.
+    if (!info.ArgBufLayout.HasTargetAddress)
+    {
+        _ASSERTE(info.Callee != NULL);
+        // TODO: enable for varargs. We need to fix the TokenLookupMap to build
+        // the proper MethodRef.
+        _ASSERTE(!info.CallSiteSig->IsVarArg());
+
+        for (COUNT_T i = 0; i < argLocals.GetCount(); i++)
+        {
+            pCode->EmitLDLOC(argLocals[i]);
+        }
+
+        if (info.CallSiteIsVirtual)
+        {
+            pCode->EmitCALLVIRT(
+                pCode->GetToken(info.Callee),
+                static_cast<int>(argLocals.GetCount()),
+                numRetVals);
+        }
+        else
+        {
+            pCode->EmitCALL(
+                pCode->GetToken(info.Callee),
+                static_cast<int>(argLocals.GetCount()),
+                numRetVals);
+        }
+    }
+    else
+    {
+        // Build the signature for the calli. TODO: Should we use the signature
```

Comment:
I think it requires it to be passed through as a `SigPointer` (right now a `MetaSig` is passed). It probably also need to be converted to a signature usable in the new module, but I believe `SigPointer::ConvertToInternalSignature` can do this for us. It might only matter once we want to support varargs.